messenger/multiple_buses.rst

```diff
@@ -0,0 +1,288 @@
+                        ),
+                    ),
+                    'messenger.bus.queries' => array(
+                        'middleware' => array(
+                            'validation',
+                        ),
+                    ),
+                    'messenger.bus.events' => array(
+                        'middleware' => array(
+                            'validation',
+                            'allow_no_handler',
+                        ),
+                    ),
+                ),
+            ),
+        ));
+
+This will generate the ``messenger.bus.commands``, ``messenger.bus.queries``
+and ``messenger.bus.events`` services that you can inject in your services.
+
+Type-hints and Auto-wiring
+--------------------------
+
+Auto-wiring is a great feature that allows you to reduce the amount of configuration
+required for your service container to be created. When using multiple buses, by default,
+the auto-wiring will not work as it won't know which bus to inject in your own services.
+
+In order to clarify this, you can use the DependencyInjection's binding capabilities
+to clarify which bus will be injected based on the argument's name:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/services.yaml
+        services:
+            _defaults:
+                # ...
+
+                bind:
+                    $commandBus: '@messenger.bus.commands'
+                    $queryBus: '@messenger.bus.queries'
+                    $eventBus: '@messenger.bus.events'
+
+    .. code-block:: xml
+
+        <!-- config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <defaults>
+                   <bind key="$commandBus" type="service" id="messenger.bus.commands" />
+                   <bind key="$queryBus" type="service" id="messenger.bus.queries" />
+                   <bind key="$eventBus" type="service" id="messenger.bus.events" />
+                </defaults>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // config/services.php
+
+        $container->bind('$commandBus', 'messenger.bus.commands');
+        $container->bind('$queryBus', 'messenger.bus.queries');
+        $container->bind('$eventBus', 'messenger.bus.events');
+
+Restrict handlers per bus & PSR-4 discovery
+-------------------------------------------
+
+By using the ``messenger.message_handler`` tag on a service, you're declaring
+it as a handler. But without more attributes, it would actually be available to
+any bus. As you don't want to dispatch a message to the wrong bus without notice,
+you can use the ``bus``  attribute with the expected bus name as value:
```

Comment:
It's a bit confusing because most users won't even be aware of this tag - it'll just be autoconfigured for them. So maybe something like:

```
By default, each handler class will be available to handle messages on *all* of your busses. To prevent
dispatching a message to the wrong bus without an error, you can restrict each handler to a specific
bus using the `messenger.message_handler` tag:
```